### ISSUE TITLE
CockroachDB Support

### ISSUE THREAD

**Issue by ansarizafar** [knex/knex#2002]
Please include support for CockroachDB [https://www.cockroachlabs.com/](https://www.cockroachlabs.com/)

**Comment by absolux** [comment]
@ansarizafar 
CockroachDB can be used with the `pg` driver [look at](https://www.cockroachlabs.com/docs/build-a-nodejs-app-with-cockroachdb.html)

**Comment by elhigu** [comment]
Support is implemented in https://github.com/tgriesser/knex/pull/1993

**Comment by nicck** [comment]
hm... I would say passing version of connected db is implemented, not CockroachDB support.

**Comment by elhigu** [comment]
@nicck is there some CockroachDB queries which are not compatible with postgresql? If so I'll reopen this as a feature request.

**Comment by nicck** [comment]
It's the opposite, PostgreSQL queries not supported by CockroachDB. I mentioned some of them in https://github.com/cockroachdb/cockroach/issues/15299#issuecomment-301061458. CockroachDB is not PostgreSQL, protocol is compatible but [CockroachDB support only subset of SQL queries](https://www.cockroachlabs.com/docs/sql-feature-support.html) which are valid in PostgreSQL.

**Comment by elhigu** [comment]
@nicck since cockroach db queries is subset of postgres queries, you should be able to use cockroach db just fine with postgres dialect. Just don't use those features that are not supported by it.

**Comment by nicck** [comment]
I can avoid using `uuid` and `json` types, but some sql queries, like [`select * for update`](https://github.com/tgriesser/knex/blob/a5f4c395297e4262aebf85534ec523509d2294e9/src/migrate/index.js#L175) and [`... table_schema = current_schema`](https://github.com/tgriesser/knex/blob/2e1a459a9e740f24b9a4647bd4da427854e551dd/src/dialects/postgres/schema/compiler.js#L22) are parts of knex as I understand and it's not possible to avoid them if you use `pg` dialect.

**Comment by elhigu** [comment]
@nicck ah right... those queries done by migration system requires some fixing. At least that select for update could be avoided by some flag which allows migration locking to throw an error instead of blocking (like it did work while ago).

So currently migration system is unusable with CockroachDB. Do you know if there are any other problems?

**Comment by thomasdavis** [comment]
I was trailing Cockroach DB yesterday and also got stuck on the migrations. Commenting as a plus one and so I can watch the thread.

**Comment by elhigu** [comment]
Meanwhile that 0.13 doesn't support it you might want to try if 0.12 does. IIRC migrations work a bit different in that version.

**Comment by kmaid** [comment]
Just tried the migration system and it won't work even with just raw sql :(.

**Comment by mhaagens** [comment]
Anyone found a solution to the migrations issue yet?
Still getting the error
```
Knex:warning - Can't take lock to run migrations: select * from "knex_migrations_lock" for update - syntax error at or near "for"
Knex:warning - If you are sure migrations are not running you can release the lock manually by deleting all the rows from migrations lock table: knex_migrations_lock
```

**Comment by mailaneel** [comment]
Temporary fix we are using for migrations to work in cockroachdb

knexfile.js

```
const Compiler = require('knex/lib/dialects/postgres/query/compiler')
const types = require('pg').types

Compiler.prototype.forUpdate = function forUpdate () {
  console.warn('table lock is not supported by cockroachdb/postgres dialect');
  return '';
};

/**
 * Required because postgres returns int as string and 
 * knex checks for isLocked are evaluating to true all the time
 */
types.setTypeParser(20, function(val) {
  return parseInt(val)
})

module.exports = {

  development: {
    client: 'postgresql',
    version: '7.2',
    debug: true,
    connection: {
      database: 'db_name',
      port: '26257',
      user: 'root',
      password: ''
    },
    pool: {
      min: 2,
      max: 10
    },
    migrations: {
      disableTransactions: true,
      tableName: 'knex_migrations'
    }
  }
};
```

**Comment by kibertoad** [comment]
Would anyone volunteer to create a failing test that would assert not triggering the undesired operation from within migrations? I consider giving this issue a shot if there is any interest in it.

**Comment by Bessonov** [comment]
@mailaneel thank you for your snippet. It worked for me.

The one thing, it can lead to errors, because the driver returns string for 64-bit integers, which are not supported in this way.

Looking forward for native support of `BigInt`: https://github.com/brianc/node-pg-types/issues/78 . Vote for it, if you want it too :o)

**Comment by Bessonov** [comment]
Related issue: https://github.com/cockroachdb/cockroach/issues/6583

**Comment by Bessonov** [comment]
I've discovered another incompatibility. CockroachDB doesn't support [savepoints](https://github.com/tgriesser/knex/issues/318). But the savepoints are a part of [Transaction Retries Protocol](https://www.cockroachlabs.com/docs/stable/transactions.html#transaction-retries). This leads to the issue, that the name of savepoint must be `cockroach_restart` and not [generated one](https://github.com/tgriesser/knex/blob/054ce36/lib/transaction.js#L19). CockroachDB can be configured to accept all names, but then you lost confidence to catch programmer errors, especially with nested transactions. Could `savepoint`-method be extended to support name parameter?

Is there any way to retry multi-query transaction with knex (without run all the code again)?

**Comment by Bessonov** [comment]
May be I did something wrong, but it turns out, that there is a more work for supporting the TRP. `tx.rollback()` leads that the transaction becomes `completed`, so there is no further actions possible. Actually, instead of sub-transaction I run raw queries:
```SQL
SAVEPOINT cockroach_restart;
-- if any 40001 error, then
ROLLBACK TO SAVEPOINT cockroach_restart;
-- run full code again
```

**Comment by kibertoad** [comment]
@Bessonov You could try extending it, if it can be implemented in a non-breaking way, we might accept the PR.

**Comment by awoods187** [comment]
Hey all, PM at Cockroach Labs here. We are adding support for savepoints in our upcoming 20.1 release. It will be in our next beta (https://www.cockroachlabs.com/docs/releases/#testing-releases). Let us know if you run into any trouble using it!

**Comment by kibertoad** [comment]
@elhigu @lorefnon @briandamaged @maximelkin Considering popularity of CockroachDB, I'm thinking of adding additional test run configuration to see how many of our tests are failing when PG driver is used to access CockroachDB (could be all of them or most of them). Might be that we do need a separate dialect (extending PG) to handle certain cases; in that case it would probably make sense to create it as a separate repo in knex project and just use as a devDependency to run tests. Thoughts?

@awoods187 If your company would consider lending any dev effort towards better support of CockroachDB in knex, that would be greatly appreciated.

**Comment by awoods187** [comment]
We generally recommend making dialects by forking the PG dialect for these types of tools. We'd be happy to answer any questions and support you along the way. We can also set up a shared slack if that'd be helpful to you.

**Comment by Bessonov** [comment]
@awoods187 thanks for the update. As I can see [the docs are not updated yet](https://www.cockroachlabs.com/docs/dev/savepoint.html). How it impacts transaction retry protocol?

@kibertoad we run knex + CRDB for a while and doesn't experienced any major problems with pg version 9.6.

ATM the only wish we have is to get rid of our custom transaction-retry-protocol implementation. There also some issues with `ALTER TABLE` for primary keys, but it works with `specificType('id', 'UUID PRIMARY KEY')`.

**Comment by awoods187** [comment]
We are currently in the process of documenting savepoints https://github.com/cockroachdb/docs/issues/5953

**Comment by vy-ton** [comment]
PM at Cockroach Labs here focused on third-party tooling.

> adding additional test run configuration to see how many of our tests are failing when PG driver is used to access CockroachDB instance
@kibertoad Any update on setting this up? Our engineering team is available to answer any questions.

**Comment by intech** [comment]
Now I have actively started using knex and have been using cockroachdb for several years, then I will try to help with tests and the implementation of syntax features.

**Comment by intech** [comment]
@kibertoad, @elhigu Is it possible to pass only postgres related tests?
I tried with env `cross-env DB=postgres npm test` but it still passes tests with sqlite.
I need to disable all unnecessary tests with the constructor and other databases, leaving only tests related to requests through one postgres driver.

**Comment by intech** [comment]
@elhigu @kibertoad I have already started to redo in a separate cockroach dialect, there will be many significant changes given the errors in the tests.
Currently 107 integration tests from postgres are failing. About 30% of them are associated with errors of the same type.

While I will analyze the errors in the tests, I can already see which are related to the features of cockroach.

What is the best way for me to proceed? Create an issue for each failed test with explanations and assign them the cockroach tag?

Can we make a separate branch in the knex repository, or can I do it in my own through PR?

All ok with `cross-env DB=cockroach KNEX_TEST_TIMEOUT=60000 npm run test:db`

**Comment by Bessonov** [comment]
Because CRDB aims to being compatible with postgres, does it make sense to address at least some incompatibilities in CRDB repo?

**Comment by intech** [comment]
@Bessonov the query builder comes up entirely from postgres with some syntax enhancements.
The schema designer is already different for me from creating / modifying tables associated with a primary key. It should be taken into account that cockroach, unlike postgres, does a lot in background tasks and many tests now when working with changing tables have not passed for this reason.

Further additional index options INVERTED / STORING / COVERING / INCLUDE which are separate for cockroach.

And the simplest thing that I corrected in the dialect is the cockroach version check. And you also need to completely change the logic associated with canceling requests, it is radically different.

**Comment by intech** [comment]
Oh yes, I forgot about the most important thing:
```
create table "primarytest" (
  "test" varchar(255),
  "test2" varchar(255) not null
);
```
at the output creates:
```
CREATE TABLE primarytest (
  test VARCHAR(255) NULL,
  test2 VARCHAR(255) NOT NULL,
  FAMILY "primary" (test, test2, rowid)
)
```
where we get the primary key an additional rowid column and again an error in the tests.
I don't think that this will be changed in the cockroach further due to the kv architecture.

**Comment by intech** [comment]
## Checklist

### Tables
- [ ] default PK auto create column [`rowid`](https://www.cockroachlabs.com/docs/stable/serial.html#use-serial-to-auto-generate-primary-keys)
- [ ] support [`INTERLEAVE IN PARENT`](https://www.cockroachlabs.com/docs/stable/interleave-in-parent)
- [ ] support [`PARTITION BY`](https://www.cockroachlabs.com/docs/stable/partitioning) differs from postgresql

### Indexes
- [ ] change PK / PK to second index - [alter table](https://www.cockroachlabs.com/docs/stable/alter-primary-key.html)
- [ ] drop PK or second index - [drop constraint](https://www.cockroachlabs.com/docs/stable/drop-constraint)
- [ ] [INVERTED](https://www.cockroachlabs.com/docs/stable/inverted-indexes) index for `jsonb`/`array`/ spatial
- [ ] `COVERING` and `INCLUDE` are aliases for [`STORING`](https://www.cockroachlabs.com/docs/stable/indexes.html#storing-columns) index
- [ ] [Hash-sharded index](https://www.cockroachlabs.com/docs/stable/create-index.html#create-a-hash-sharded-secondary-index)
- [ ] (**v20.2+**) [spatial indexes](https://www.cockroachlabs.com/docs/v20.2/spatial-indexes)

### Columns
- [ ] default value by `unique_rowid()` for type `int`
- [ ] default value by `gen_random_uuid()` for type `uuid`
- [ ] [`CHECK`](https://www.cockroachlabs.com/docs/stable/check) constraint
- [ ] Unsupported type `enum` or support `CREATE TYPE ... AS ENUM`

### Queries / Transactions
- [ ] support [`UPSERT`](https://www.cockroachlabs.com/docs/stable/upsert)
- [ ] support select and transaction with [`AS OF SYSTEM TIME <time>`](https://www.cockroachlabs.com/docs/stable/as-of-system-time.html)
- [ ] `where` with time using [cost-based optimizer](https://www.cockroachlabs.com/docs/stable/known-limitations.html#filtering-by-now-results-in-a-full-table-scan) without full table scan
- [ ] column name from an outer column inside a subquery as [`?column?`](https://www.cockroachlabs.com/docs/stable/known-limitations.html#column-name-from-an-outer-column-inside-a-subquery-differs-from-postgresql)
- [ ] Limitation of [schema changes within transactions](https://www.cockroachlabs.com/docs/stable/known-limitations.html#schema-changes-within-transactions)

### Dialect
- [ ] version check
- [ ] support BigInt
- [ ] [cancel query](https://www.cockroachlabs.com/docs/stable/manage-long-running-queries)
- [ ] unsupported stream
- [ ] add [functions](https://www.cockroachlabs.com/docs/stable/functions-and-operators.html) in `FunctionHelper` and refactoring for difference dialect
- [ ] **(v20.2+)** `for share` / `skip locked` / `nowait` https://github.com/cockroachdb/cockroach/issues/40476

**Comment by intech** [comment]
@vy-ton @awoods187 please, check the list and add information on version 20.2 what other changes will be related to postgres compatibility support

**Comment by vy-ton** [comment]
Awesome to see things happening here!

In 20.2, CockroachDB will add the following

- [ ] Postgres [schemas](https://www.postgresql.org/docs/current/ddl-schemas.html)
- [ ] Enums
- [ ] Materialized views
- [ ] Partial indexes

**Comment by intech** [comment]
@vy-ton Thank you for your prompt reply!

**Comment by kibertoad** [comment]
@intech Separate branch in knex repo makes sense, I'll create one. I can add CockroachDB tag for the repo.
Would you like to be an official maintainer of CockroachDB support in knex, btw?

**Comment by intech** [comment]
@kibertoad yes, I only use knex with cockroach, it will be easiest for me to provide support from my own experience.

**Comment by kibertoad** [comment]
@intech Any progress on this? Doing major refactorings of knex right now, if there any any internal changes needed to support some of the CockroachDB pecularities, would be glad to help.

**Comment by intech** [comment]
@kibertoad Yes, work is underway. I studied the codebase and just realized that it would be better not to do it in the same codebase following the example of other drivers, but to rewrite it into new classes with a separate main column constructor, column modifiers, rules for building schema indexes, because in cockroach the logic of the optimizer is different.
This is great news because I have some driver refactoring requests.
I'll write more later. Thank you.

**Comment by kibertoad** [comment]
@intech I agree, existing driver structure is not necessarily the best overall, so clean-slate implementation might be a good idea all-around.
Yup, please keep me in the loop, will do do necessary adjustments ASAP.

**Comment by vy-ton** [comment]
@intech @kibertoad Hi, Product Manager at Cockroach Labs here. 

I would love to discuss how the work @intech started is going and how the CockroachDB team could help with these efforts?

**Comment by intech** [comment]
@vy-ton Sorry for the long awaited response.
Right now I have added the previously missed `WINDOW` functions to my checklist.
As for the progress of the work, I have been busy for a while, but I will return to the task soon. For me it is very relevant in my current work.

**Comment by vy-ton** [comment]
@intech That sounds great!
I can also set up a video call with our engineers? They can provide insights into the remaining failed tests.

**Comment by jahudka** [comment]
Hi, thanks everybody for your work! I'm mostly writing to add a "plus one" to show interest in this feature and to get notifications about further updates, but I also have a question which seems relevant enough to ask here: will this driver support specifying multiple hosts in the connection URL, similarly to what `libpq` allows?

TL;DR - In `libpq`-based clients one can specify multiple `host:port` pairs in the connection URL separated by a comma, so e.g. `postgresql://user@host1:5432,host2:5432,host3:5432/dbname`. As I understand it, `libpq` should then try connecting to the hosts in the order they're specified and use the first one that succeeds. I'm not sure my logic is undeniable here, but I think that having an app always connect to the same hardcoded Cockroach node kind of defeats the purpose of having a distributed database, and having to rely on DNS to resolve a hostname to a running node adds both complexity and an unnecessary dependency.

The `pg` driver doesn't seem to support this yet, but `pg-native` does (since it uses `libpq` bindings under the hood), and `pg` can wrap `pg-native` and re-expose it with (mostly) the same API, so in theory, if this driver is built on top of `pg`, supporting this feature could be as easy as `require('pg').native` instead of `require('pg')`. This would kill a couple of features that `pg-native` can't support, like query streaming, but it might not be an unacceptable tradeoff for some use cases.

**Comment by kibertoad** [comment]
@jahudka There is mostly complete support for pg-native: #4327

**Comment by jahudka** [comment]
@kibertoad yeah I know, but if I understand correctly, the `pgnative` driver, when finished, will be compatible with Postgres, but not with Cockroach (in the sense that it won't solve the problems that @intech's Cockroach driver discussed here solves), right? So I'll then have either the option to use the Cockroach driver, but only connect to a single configured node, or to use `pgnative`, but maybe some stuff like migrations won't work.. if that's the case then that might prevent me from using `pgnative` completely, because what I'd _really_ like to use is MikroORM, which is built on top of Knex, and so I might not be able to decide to avoid some Knex features that break when `pgnative` is used with CockroachDb..

**Comment by intech** [comment]
@jahudka don't worry about streaming, cockroach doesn't support it. 

You are right, pg does not have the ability to pass a list of nodes, and I agree that this should be added to the pg package. The problem is in the architecture of different databases using the same postgres dialect, that not all of them work in master-master mode and failover will be different.
We need to take as an example the configuration used in redis with priorities and master/slave roles.

**Comment by vy-ton** [comment]
@intech If you have time over the next few months to work on the CockroachDB support, I would like to discuss a sponsorship opportunity.

**Comment by intech** [comment]
> @intech If you have time over the next few months to work on the CockroachDB support, I would like to discuss a sponsorship opportunity.

Wow, this is unexpected! I have interests to be closer to Cockroach Labs, let's continue communicating by [email](mailto:ru31337@gmail.com) (we have already communicated with Andy, Rafi and Jesse earlier).

**Comment by kibertoad** [comment]
@intech I've started working on CockroachDB as a Knex dialect. Are there any insights or changes you had to do so far, that you could share?

**Comment by intech** [comment]
@kibertoad After refactoring Knex itself is not yet. Make a cockroachdb branch with base dialect from postgres and we will work on it together.

**Comment by intech** [comment]
Saw already in https://github.com/knex/knex/tree/feat/cockroachdb!

**Comment by rustyconover** [comment]
Any luck yet with migrations and cockroachdb?

**Comment by kibertoad** [comment]
Working on it!

**Comment by kibertoad** [comment]
@intech Rudimentary CockroachDB landed in master already. Not all tests pass yet, but core functionality should work already. Any chance you could take a look, and maybe you have any code for the driver itself on top of what we currently have that you could share?

**Comment by kibertoad** [comment]
This is coming in 0.95.12, most likely late October!

**Comment by rafiss** [comment]
I believe this can be closed now, correct?

**Comment by intech** [comment]
> I believe this can be closed now, correct?

Yes, support can be tracked by the [CockroachDB label](https://github.com/knex/knex/issues?q=is%3Aopen+is%3Aissue+label%3ACockroachDB) and [wiki page](https://github.com/knex/knex/wiki/CockroachDB-Support)

**Comment by jmsunseri** [comment]
I still need to add `disableTransactions: true,` in my knexfile in order to be able to run migrate:latest or i get an error like this.

```
COMMIT; - restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh): "sql txn" meta={id=da05773f key=/Tenant/1929/NamespaceTable/30/1/87/29/"projects"/4/1 pri=0.01368187 epo=0 ts=1642501152.147079958,1 min=1642501146.536358536,0 seq=160} lock=true stat=PENDING rts=1642501146.536358536,0 wto=false gul=1642501147.036358536,0
error: COMMIT; - restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE - failed preemptive refresh): "sql txn" meta={id=da05773f key=/Tenant/1929/NamespaceTable/30/1/87/29/"projects"/4/1 pri=0.01368187 epo=0 ts=1642501152.147079958,1 min=1642501146.536358536,0 seq=160} lock=true stat=PENDING rts=1642501146.536358536,0 wto=false gul=1642501147.036358536,0
    at Parser.parseErrorMessage (/Users/jmsunseri/dev/code/flash-card-epub-builder/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/Users/jmsunseri/dev/code/flash-card-epub-builder/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/Users/jmsunseri/dev/code/flash-card-epub-builder/node_modules/pg-protocol/dist/parser.js:39:38)
    at TLSSocket.<anonymous> (/Users/jmsunseri/dev/code/flash-card-epub-builder/node_modules/pg-protocol/dist/index.js:11:42)
    at TLSSocket.emit (events.js:400:28)
    at addChunk (internal/streams/readable.js:293:12)
    at readableAddChunk (internal/streams/readable.js:267:9)
    at TLSSocket.Readable.push (internal/streams/readable.js:206:10)
    at TLSWrap.onStreamRead (internal/stream_base_commons.js:188:23)
```

 migrate:up works fine.

**Comment by kibertoad** [comment]
@jmsunseri Which knex version?

**Comment by jmsunseri** [comment]
"knex": "^0.95.15"


Sincerely,
Justin Sunseri

**Comment by dedisuryadi** [comment]
Same here, migration running fine on docker with dev mode on, but got error when migrating to remote machine, ```disableTransactions: true``` saved me, thank you @jmsunseri 

```
migration failed with error: select * from "knex_migrations_lock" - restart transaction: TransactionRetryWithProtoRefreshError: ReadWithinUncertaintyIntervalError: read at time 1669183964.437105736,0 encountered previous write with future timestamp 1669183964.613185698,0 within uncertainty interval `t <= (local=1669183964.756284634,0, global=1669183964.937105736,0)`; observed timestamps: [{1 1669183964.437105736,0} {2 1669183964.756284634,0}]: "sql txn" meta={id=dae4d74f pri=0.00170548 epo=0 ts=1669183964.437105736,0 min=1669183964.437105736,0 seq=0} lock=false stat=PENDING rts=1669183964.437105736,0 wto=false gul=1669183964.937105736,0
select * from "knex_migrations_lock" - restart transaction: TransactionRetryWithProtoRefreshError: ReadWithinUncertaintyIntervalError: read at time 1669183964.437105736,0 encountered previous write with future timestamp 1669183964.613185698,0 within uncertainty interval `t <= (local=1669183964.756284634,0, global=1669183964.937105736,0)`; observed timestamps: [{1 1669183964.437105736,0} {2 1669183964.756284634,0}]: "sql txn" meta={id=dae4d74f pri=0.00170548 epo=0 ts=1669183964.437105736,0 min=1669183964.437105736,0 seq=0} lock=false stat=PENDING rts=1669183964.437105736,0 wto=false gul=1669183964.937105736,0
error: select * from "knex_migrations_lock" - restart transaction: TransactionRetryWithProtoRefreshError: ReadWithinUncertaintyIntervalError: read at time 1669183964.437105736,0 encountered previous write with future timestamp 1669183964.613185698,0 within uncertainty interval `t <= (local=1669183964.756284634,0, global=1669183964.937105736,0)`; observed timestamps: [{1 1669183964.437105736,0} {2 1669183964.756284634,0}]: "sql txn" meta={id=dae4d74f pri=0.00170548 epo=0 ts=1669183964.437105736,0 min=1669183964.437105736,0 seq=0} lock=false stat=PENDING rts=1669183964.437105736,0 wto=false gul=1669183964.937105736,0
    at Parser.parseErrorMessage (/Users/REDACTED/node_modules/pg-protocol/dist/parser.js:287:98)
    at Parser.handlePacket (/Users/REDACTED/node_modules/pg-protocol/dist/parser.js:126:29)
    at Parser.parse (/Users/REDACTED/node_modules/pg-protocol/dist/parser.js:39:38)
    at Socket.<anonymous> (/Users/REDACTED/node_modules/pg-protocol/dist/index.js:11:42)
    at Socket.emit (node:events:390:28)
    at addChunk (node:internal/streams/readable:315:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Socket.Readable.push (node:internal/streams/readable:228:10)
    at TCP.onStreamRead (node:internal/stream_base_commons:199:23)
```
